### PR TITLE
Improve rids counter

### PR DESCRIPTION
**Pull request by Nicogp**
|Related issue|
|---|
|#6112|

## Description
This PR add the functionality to close a rids fp after a seted time. By default this time is 300 seconds and it´s configurable in the remote section of ossec.conf with tag "rids_closing_time".

## Configuration options
```
<remote>
   <rids_closing_time>10m</rids_closing_time>
</remote>
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
- [x] Source installation
- [x] Source upgrade
- [x] Review logs syntax and correct language

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)

### CODE REVIEW

File: src/remoted/secure.c [hunk 1]
```diff
@@ -271,6 +276,47 @@ void * rem_keyupdate_main(__attribute__((unused)) void * args) {
...
+                mdebug2("Checking rids_node of agent %s.", first_node_key->id);
+                if ((now - seconds) > first_node_key->updating_time) {
+                    first_node_key = (keyentry *)linked_queue_pop_ex(keys->opened_fp_queue);
+                    mdebug2("Pop rids_node of agent %s.", first_node_key->id);
+                    if (first_node_key->fp != NULL) {
+                        mdebug2("Closing rids for agent %s.", first_node_key->id);
+                        fclose(first_node_key->fp);
+                    }
+                    first_node_key->updating_time = 0;
+                    first_node_key->fp = NULL;
```
Comment (TomasTurina) [inline src/remoted/secure.c:307]: Move this line inside the if above.

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/os_crypto/shared/msgs.c [hunk 2]
```diff
@@ -393,6 +411,7 @@ int ReadSecMSG(keystore *keys, char *buffer, char *cleartext, int id, unsigned i
                 StoreCounter(keys, id, msg_global, msg_local);
                 rcv_count = 0;
             }
+            w_mutex_unlock(&keys->keyentries[id]->mutex);
```
Comment (TomasTurina) [inline src/os_crypto/shared/msgs.c:414]: Close the mutex after increasing `rcv_count` to also protect that variable.

File: src/os_crypto/shared/msgs.c [hunk 3]
```diff
@@ -373,15 +391,15 @@ int ReadSecMSG(keystore *keys, char *buffer, char *cleartext, int id, unsigned i
...
```
Comment (TomasTurina) [inline src/os_crypto/shared/msgs.c:394]: Close the mutex after increasing `rcv_count` to also protect that variable.

File: src/os_crypto/shared/msgs.c [hunk 4]
```diff
@@ -407,10 +426,11 @@ int ReadSecMSG(keystore *keys, char *buffer, char *cleartext, int id, unsigned i
                 msg_local,
                 keys->keyentries[id]->global,
                 keys->keyentries[id]->local);
-
+            w_mutex_unlock(&keys->keyentries[id]->mutex);
             merror(ENCTIME_ERROR, keys->keyentries[id]->name);
             return KS_RIDS;
         }
+        w_mutex_unlock(&keys->keyentries[id]->mutex);
```
Comment (TomasTurina) [inline src/os_crypto/shared/msgs.c:433]: This is not necessary, you are already unlocking the mutex at the end of the function.

**Review by TomasTurina — CHANGES_REQUESTED** [review]

File: src/remoted/secure.c [hunk 5]
```diff
@@ -272,6 +277,48 @@ void * rem_keyupdate_main(__attribute__((unused)) void * args) {
...
+    keystore * keys = (keystore *)args;
+    int seconds;
+    int flag;
+
+    mdebug1("Rids closer thread started.");
+    seconds = logr.rids_closing_time;
+
+    while (1) {
+        sleep(seconds);
+        key_lock_write();
```
Comment (nicolaspapp) [inline src/remoted/secure.c:291]: This looks expensive and rather unnecessary. In my understaing:
- key_lock_read(): Some thread that is reading the keystore
- keyentries[i]->mutex: Mutual exclusion of that key entry. Can modifcate keyentry information in the process
- key_lock_write(): Block all reads to the keystore. Possible needed when keystore is reloaded or new keys are added

In my opinion only read lock is needed here. Also need to take the keyentry mutex before evaulating the file pointer

File: src/remoted/secure.c [hunk 6]
```diff
@@ -272,6 +277,48 @@ void * rem_keyupdate_main(__attribute__((unused)) void * args) {
...
+        flag = 1;
+        while (flag) {
+            w_linked_queue_node_t * first_node = keys->opened_fp_queue->first;
+            mdebug2("Opened rids queue size: %d", keys->opened_fp_queue->elements);
+            if (first_node) {
+                int now = time(0);
+                keyentry * first_node_key = (keyentry *)first_node->data;
+                mdebug2("Checking rids_node of agent %s.", first_node_key->id);
+                if ((now - seconds) > first_node_key->updating_time) {
+                    first_node_key = (keyentry *)linked_queue_pop_ex(keys->opened_fp_queue);
```
Comment (nicolaspapp) [inline src/remoted/secure.c:301]: mutex lock here

File: src/remoted/secure.c [hunk 7]
```diff
@@ -272,6 +277,48 @@ void * rem_keyupdate_main(__attribute__((unused)) void * args) {
...
+                    mdebug2("Pop rids_node of agent %s.", first_node_key->id);
+                    if (first_node_key->fp != NULL) {
+                        mdebug2("Closing rids for agent %s.", first_node_key->id);
+                        fclose(first_node_key->fp);
+                        first_node_key->fp = NULL;
+                    }
+                    first_node_key->updating_time = 0;
+                    first_node_key->rids_node = NULL;
```
Comment (nicolaspapp) [inline src/remoted/secure.c:309]: mutex_unlock

**Review by nicolaspapp — CHANGES_REQUESTED** [review]

**Review by TomasTurina — APPROVED** [review]

File: src/unit_tests/os_crypto/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+# This program is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public
+# License (version 2) as published by the FSF - Free Software
+# Foundation.
+
+cmake_minimum_required(VERSION 3.10)
+
+if(${TARGET} STREQUAL "server")
+    add_subdirectory(shared)
+endif()
```
Comment (nicolaspapp) [inline src/unit_tests/os_crypto/CMakeLists.txt:12]: Add EOF

File: src/unit_tests/os_crypto/shared/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        )
+    endif()
+    if(NOT test_flags STREQUAL " ")
+        target_link_libraries(
+            ${test_name}
+            ${test_flags}
+        )
+    endif()
+    add_test(NAME ${test_name} COMMAND ${test_name})
+endforeach()
```
Comment (nicolaspapp) [inline src/unit_tests/os_crypto/shared/CMakeLists.txt:34]: Add EOF

File: src/unit_tests/os_crypto/shared/test_msgs.c [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+{
+    const struct CMUnitTest tests[] = {
+        // Tests StoreCounter
+        cmocka_unit_test_setup_teardown(test_StoreCounter_updating_rids, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_StoreCounter_pushing_rids, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_StoreCounter_pushing_rids_fp_null, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_StoreCounter_fail_first_open, setup_config, teardown_config)
+        };
+    return cmocka_run_group_tests(tests, NULL, NULL);
+}
```
Comment (nicolaspapp) [inline src/unit_tests/os_crypto/shared/test_msgs.c:287]: Add EOF

File: src/unit_tests/remoted/test_secure.c [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+        // Tests close_fp_main
+        cmocka_unit_test_setup_teardown(test_close_fp_main_queue_empty, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_close_fp_main_first_node_no_close_first, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_close_fp_main_close_first, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_close_fp_main_close_first_queue_2, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_close_fp_main_close_first_queue_2_close_2, setup_config, teardown_config),
+        cmocka_unit_test_setup_teardown(test_close_fp_main_close_fp_null, setup_config, teardown_config)
+        };
+    return cmocka_run_group_tests(tests, NULL, NULL);
+}
```
Comment (nicolaspapp) [inline src/unit_tests/remoted/test_secure.c:358]: Add EOF

**Review by nicolaspapp — APPROVED** [review]

**Review by TomasTurina — APPROVED** [review]

**Pull request merged**